src/cluster_legacy.c

```diff
@@ -1470,15 +1470,28 @@ int clusterNodeRemoveSlave(clusterNode *master, clusterNode *slave) {
 }
 
 int clusterNodeAddSlave(clusterNode *master, clusterNode *slave) {
-    int j;
+    int j, pos = -1;
 
     /* If it's already a slave, don't add it again. */
-    for (j = 0; j < master->numslaves; j++)
+    for (j = 0; j < master->numslaves; j++) {
         if (master->slaves[j] == slave) return C_ERR;
-    master->slaves = zrealloc(master->slaves,
-        sizeof(clusterNode*)*(master->numslaves+1));
-    master->slaves[master->numslaves] = slave;
+        /* Find lexicographic order index for the slave. */
+        if (pos == -1 && strcasecmp(slave->name, master->slaves[j]->name) > 0) {
+            pos = j;
+        }
+    }
+    /* If no slave(s) are found or it's lexicographically smallest. */
+    if (pos == -1) pos = 0;
+
     master->numslaves++;
+    master->slaves = zrealloc(master->slaves, sizeof(clusterNode*)*(master->numslaves));
+    
+    /* Copy all remaining nodes to the right. */
+    int remaining_slaves = (master->numslaves - pos) - 1;
+    if (remaining_slaves > 0) {
+        memmove(master->slaves+(pos+1), master->slaves+pos, (sizeof(*master->slaves) * remaining_slaves));
```

Comment:
This originally seemed right to me, but I guess maybe we also consider:
```
int clusterNodeNameComparator(const void *arg1, const void *arg2) {
    return strncasecmp(((clusterNode *) arg1->name, (clusterNode *) arg2->name, CLUSTER_NAMELEN);
}

qsort(master->slaves, master->numslaves, sizeof(clusterNode *), clusterNodeNameComparator);
```

It seems simpler, and we really don't care about the extra time complexity here given how unlikely it will be executed. Up to you though.

Reply:
Was very much influenced by our interaction. I also think `qsort` is much cleaner/less error prone.